.github/workflows/build-and-push-docker-images.yml

```
@@ -39,9 +39,40 @@ jobs:
     uses: ./.github/workflows/build-all-binaries.yml
     secrets: inherit
 
-  build-and-push:
+  # Get the common things so we're not doing slow operations
+  # for every matrix item
+  download-artifacts-for-docker-build:
     needs: build-binaries
     runs-on: ["randamu-self-hosted-default"]
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v5
+        with:
+          submodules: 'recursive'
+
+      - name: Download pre-built binaries
+        uses: actions/download-artifact@v4
+        with:
+          name: binaries-${{ github.sha }}
+          path: target/release
+
+      - name: Upload binaries for docker jobs
+        uses: actions/upload-artifact@v4
+        with:
+          name: binaries-for-docker-${{ github.sha }}
+          path: target/release/
+          retention-days: 1
```

overwrite probably makes sense.
```suggestion
          retention-days: 1
          overwrite: true
```